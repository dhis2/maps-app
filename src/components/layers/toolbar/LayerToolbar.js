import React from 'react';
import PropTypes from 'prop-types';
import { Toolbar, ToolbarGroup } from 'material-ui/Toolbar';
import IconButton from 'material-ui/IconButton';
import IconMenu from 'material-ui/IconMenu';
import { SvgIcon } from '@dhis2/d2-ui-core';
import OpacitySlider from './OpacitySlider';
import DownloadMenu from './DownloadMenu';
import OpenAsMenu from './OpenAsMenu';

const styles = {
    toolbar: {
        backgroundColor: '#eee',
        height: 32,
        padding: '0 8px',
    },
    button: {
        float: 'left',
        padding: 4,
        width: 32,
        height: 32,
    },
    moreButton: {
        float: 'right',
        padding: 4,
        width: 32,
        height: 32,
        marginRight: -7,
        marginLeft: -5,
    },
    menuList: {
        paddingTop: 0,
        paddingBottom: 0,
    },
};

const OverlayToolbar = ({
    layer,
    onEdit,
    onRemove,
    toggleDataTable,
    onOpacityChange,
}) => (
    <Toolbar style={styles.toolbar}>
        <ToolbarGroup>
            {onEdit &&
                layer.type !== 'external' && (
                    <IconButton
                        onClick={() => onEdit(layer)}
                        tooltip="Edit"
                        tooltipPosition="top-center"
                        style={styles.button}
                    >
                        <SvgIcon icon="Create" />
                    </IconButton>
                )}
            {(layer.layer === 'thematic' ||
                layer.layer === 'boundary' ||
                layer.layer === 'facility') && (
                <IconButton
                    onClick={() => toggleDataTable(layer.id)}
                    tooltip="Data table"
                    tooltipPosition="top-center"
                    style={styles.button}
                >
                    <SvgIcon icon="ViewList" />
                </IconButton>
            )}

            <OpacitySlider
                {...layer}
                onChange={opacity => onOpacityChange(layer.id, opacity)}
            />
        </ToolbarGroup>

        <ToolbarGroup>
            {onRemove && (
                <IconButton
                    onClick={onRemove}
                    tooltip="Delete"
                    tooltipPosition="top-center"
                    style={styles.button}
                >
                    <SvgIcon icon="Delete" />
                </IconButton>
            )}

            {true === false && ( //    TODO
                <IconMenu
                    iconButtonElement={
                        <IconButton
                            tooltip="More"
                            tooltipPosition="top-center"
                            style={styles.moreButton}
                        >
                            <SvgIcon icon="MoreVert" />
                        </IconButton>
                    }
                    listStyle={styles.menuList}
                >
                    <OpenAsMenu {...layer} />
                    <DownloadMenu {...layer} />
                </IconMenu>
            )}
        </ToolbarGroup>
    </Toolbar>
);

OverlayToolbar.propTypes = {
    layer: PropTypes.object,
    onEdit: PropTypes.func,
    onRemove: PropTypes.func,
    onDataTableShow: PropTypes.func,
    onOpacityChange: PropTypes.func,
};

export default OverlayToolbar;
