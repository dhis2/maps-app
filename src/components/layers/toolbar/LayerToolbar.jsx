import i18n from '@dhis2/d2-i18n'
import { Tooltip, IconEdit24, IconView24, IconViewOff24 } from '@dhis2/ui'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import { IconButton } from '../../core/index.js'
import LayerToolbarMoreMenu from './LayerToolbarMoreMenu.jsx'
import OpacitySlider from './OpacitySlider.jsx'
import styles from './styles/LayerToolbar.module.css'

const LayerToolbar = ({
    hasOpacity = true,
    opacity = 1,
    isVisible,
    onOpacityChange,
    toggleLayerVisibility,
    hasError,
    ...expansionMenuProps
}) => {
    const onEdit = expansionMenuProps.onEdit

    return (
        <div className={styles.toolbar} data-test="layertoolbar">
            {onEdit && (
                <IconButton
                    tooltip={i18n.t('Edit')}
                    onClick={onEdit}
                    className={styles.editButton}
                    dataTest="layer-edit-button"
                >
                    <IconEdit24 />
                </IconButton>
            )}
            <IconButton
                tooltip={!hasError ? i18n.t('Toggle visibility') : null}
                onClick={toggleLayerVisibility}
                className={cx({
                    visible: isVisible, // for cypress testing only
                    notvisible: !isVisible, // for cypress testing only
                })}
                dataTest="visibilitybutton"
                disabled={hasError}
            >
                {isVisible ? <IconView24 /> : <IconViewOff24 />}
            </IconButton>

            <div className={styles.sliderContainer}>
                <Tooltip
                    content={
                        hasError
                            ? i18n.t('Layer is invalid')
                            : i18n.t('Set layer opacity')
                    }
                    disabled={hasError}
                >
                    <OpacitySlider
                        opacity={opacity}
                        disabled={hasError || !isVisible || !hasOpacity}
                        onChange={onOpacityChange}
                    />
                </Tooltip>
            </div>
            <div className={styles.menuButton}>
                <LayerToolbarMoreMenu
                    hasError={hasError}
                    {...expansionMenuProps}
                />
            </div>
        </div>
    )
}

LayerToolbar.propTypes = {
    toggleLayerVisibility: PropTypes.func.isRequired,
    onOpacityChange: PropTypes.func.isRequired,
    hasError: PropTypes.bool,
    hasOpacity: PropTypes.bool,
    isVisible: PropTypes.bool,
    opacity: PropTypes.number,
    onEdit: PropTypes.func,
}

export default LayerToolbar
