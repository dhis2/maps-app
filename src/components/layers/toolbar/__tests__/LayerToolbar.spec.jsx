import { render, fireEvent } from '@testing-library/react'
import React from 'react'
import LayerToolbar from '../LayerToolbar.jsx'

/* eslint-disable react/prop-types */
jest.mock('@dhis2/ui', () => {
    const originalModule = jest.requireActual('@dhis2/ui')

    return {
        __esModule: true,
        ...originalModule,
        Tooltip: function Tooltip({ children }) {
            return <div>{children}</div>
        },
        IconView24: function IconView24() {
            return <div>IconView24</div>
        },
        IconViewOff24: function IconViewOff24() {
            return <div>IconViewOff24</div>
        },
    }
})

jest.mock('../LayerToolbarMoreMenu.jsx', () => {
    return function LayerToolbarMoreMenu() {
        return <div>LayerToolbarMoreMenu</div>
    }
})
/* eslint-enable react/prop-types */

const props = {
    opacity: 0,
    isVisible: true,
    toggleLayerVisibility: jest.fn(),
    onOpacityChange: jest.fn(),
}

describe('LayerToolbar', () => {
    it('Should render only a visibility toggle and opacity slider', () => {
        const { container } = render(
            <LayerToolbar {...props} isVisible={true} />
        )
        expect(container).toMatchSnapshot()
    })

    it('Should show SvgViewOff24 when not visible', () => {
        const { container } = render(
            <LayerToolbar {...props} isVisible={false} />
        )
        expect(container).toMatchSnapshot()
    })

    it('Should call toggleLayerVisibility callback on button press', async () => {
        const toggleVisibleFn = jest.fn()
        const { container } = render(
            <LayerToolbar {...props} toggleLayerVisibility={toggleVisibleFn} />
        )

        await fireEvent.click(
            container.querySelector('[data-test="visibilitybutton"]')
        )
        expect(toggleVisibleFn).toHaveBeenCalledTimes(1)
    })

    it('Should render edit button', () => {
        const { container } = render(
            <LayerToolbar {...props} onEdit={jest.fn()} />
        )
        expect(container).toMatchSnapshot()
    })

    it('Should call onEdit callback on button press', async () => {
        const toggleVisibleFn = jest.fn()
        const editFn = jest.fn()

        const { container } = render(
            <LayerToolbar
                {...props}
                onEdit={editFn}
                toggleLayerVisibility={toggleVisibleFn}
            />
        )

        await fireEvent.click(
            container.querySelector('[data-test="layer-edit-button"]')
        )
        expect(editFn).toHaveBeenCalledTimes(1)
        expect(toggleVisibleFn).not.toHaveBeenCalled()

        await fireEvent.click(
            container.querySelector('[data-test="visibilitybutton"]')
        )
        expect(toggleVisibleFn).toHaveBeenCalledTimes(1)
        expect(editFn).toHaveBeenCalledTimes(1)
    })
})
