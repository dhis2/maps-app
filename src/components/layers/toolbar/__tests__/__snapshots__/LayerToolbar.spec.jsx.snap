// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerToolbar Should render edit button 1`] = `
<div>
  <div
    class="toolbar"
    data-test="layertoolbar"
  >
    <button
      class="iconButton editButton"
      data-test="layer-edit-button"
    >
      <div>
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17.287 4.297l.127.117 2.172 2.172a2 2 0 01.117 2.701l-.117.127L9 20H4v-5L14.586 4.414a2 2 0 012.701-.117zM16 5.828l-9.586 9.585 2.171 2.171L18.172 8z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
    <button
      class="iconButton visible"
      data-test="visibilitybutton"
    >
      <div>
        <div>
          IconView24
        </div>
      </div>
    </button>
    <div
      class="sliderContainer"
    >
      <div>
        <div
          class="container"
        >
          <input
            class="slider"
            max="1"
            min="0"
            step="0.01"
            type="range"
            value="0"
          />
        </div>
      </div>
    </div>
    <div
      class="menuButton"
    >
      <div>
        LayerToolbarMoreMenu
      </div>
    </div>
  </div>
</div>
`;

exports[`LayerToolbar Should render only a visibility toggle and opacity slider 1`] = `
<div>
  <div
    class="toolbar"
    data-test="layertoolbar"
  >
    <button
      class="iconButton visible"
      data-test="visibilitybutton"
    >
      <div>
        <div>
          IconView24
        </div>
      </div>
    </button>
    <div
      class="sliderContainer"
    >
      <div>
        <div
          class="container"
        >
          <input
            class="slider"
            max="1"
            min="0"
            step="0.01"
            type="range"
            value="0"
          />
        </div>
      </div>
    </div>
    <div
      class="menuButton"
    >
      <div>
        LayerToolbarMoreMenu
      </div>
    </div>
  </div>
</div>
`;

exports[`LayerToolbar Should show SvgViewOff24 when not visible 1`] = `
<div>
  <div
    class="toolbar"
    data-test="layertoolbar"
  >
    <button
      class="iconButton notvisible"
      data-test="visibilitybutton"
    >
      <div>
        <div>
          IconViewOff24
        </div>
      </div>
    </button>
    <div
      class="sliderContainer"
    >
      <div>
        <div
          class="container"
        >
          <input
            class="slider"
            disabled=""
            max="1"
            min="0"
            step="0.01"
            type="range"
            value="0"
          />
        </div>
      </div>
    </div>
    <div
      class="menuButton"
    >
      <div>
        LayerToolbarMoreMenu
      </div>
    </div>
  </div>
</div>
`;
